src/main/java/randoop/generation/Bloodhound.java

```diff
@@ -0,0 +1,235 @@
+   */
+  private final double p = 0.5;
+
+  /**
+   * Branch coverage is recomputed after this many successful invocations (= this many new tests
+   * were generated).
+   */
+  private final int branchCoverageInterval = 100;
+
+  /** The total number of successful invocations of all the methods under test. */
+  private int totalSuccessfulInvocations = 0;
+
+  /**
+   * Maximum number of times any method under test has been successfully invoked. This value is
+   * initialized to 1 because it is used as the denominator of a division in computing a method's
+   * weight. The name is from the GRT paper, which calls this quantity "maxSucc(M)".
+   */
+  private int maxSuccM = 1;
+
+  /**
+   * Initialize Bloodhound.
+   *
+   * @param operations list of operations under test
+   */
+  public Bloodhound(List<TypedOperation> operations, CoverageTracker coverageTracker) {
+    this.operationSimpleList = new SimpleArrayList<>(operations);
+    this.coverageTracker = coverageTracker;
+    // Compute an initial weight for all methods under test. We also initialize the uncovered ratio
+    // value of all methods under test by updating branch coverage information. The weights for all methods may not be uniform
+    // in cases where we have methods with "zero" branches and methods with non-"zero" branches.
+    updateBranchCoverageMaybe();
+    updateWeightsForAllOperations();
+  }
+
+  /**
+   * Selects a method under test for the {@link ForwardGenerator} to use to construct a new
+   * sequence. A method under test is randomly selected with a weight probability.
+   *
+   * <p>Branch coverage information, which is used to compute weights for methods under test, is
+   * updated at every {@code branchCoverageInterval}'th call of this method. The selection count for
+   * the selected method is incremented in the {@code methodSelectionCounts} map. Finally, the
+   * weight of the selected method is recomputed.
+   *
+   * @return the chosen {@code TypedOperation} for the new sequence
+   */
+  @Override
+  public TypedOperation selectOperation() {
+    // Periodically collect branch coverage and recompute weights for methods under test.
+    updateBranchCoverageMaybe();
+
+    // Make a random, weighted choice for the next method.
+    TypedOperation selectedOperation =
+        Randomness.randomMemberWeighted(operationSimpleList, methodWeights);
+
+    // Update the selected method's selection count and recompute its weight.
+    incrementInMap(methodSelectionCounts, selectedOperation);
+    updateWeightForOperation(selectedOperation);
+
+    return selectedOperation;
+  }
+
+  /**
+   * At every {@code branchCoverageInterval}'th call of this method, the branch coverage information
+   * for all methods under test is re-summarized and weights for all methods under test are
+   * recomputed.
+   */
+  private void updateBranchCoverageMaybe() {
+    if (totalSuccessfulInvocations % branchCoverageInterval == 0) {
+      methodSelectionCounts.clear();
```

Comment:
I think that it could be, however `methodSelectionCounts` belongs to Bloodhound and `summarizeCoverageInformation` belongs to CoverageTracker. To clear the map, would it be a good idea to pass the methodSelectionCounts map into summarizeCoverageInformation?

Reply:
Alternatively, did you mean to abstract out the two lines into a separate method that would both clear the methodselectionscount map and call summarizeCoverageInformation on the coverageTracker instance?